### PR TITLE
gitrepo: initial implementaiton

### PR DESCRIPTION
**Pull request by kopiczko**
Towards: https://github.com/giantswarm/giantswarm/issues/6844

This PR implements the core logic of this library, which is to determine a project version based on a git reference like the `architect project version` command.
The reason we need this is because we need to be able to determine version in other places like `opsctl`, so we need a re-usable piece of code.

### CODE REVIEW

File: pkg/gitrepo/repo.go [hunk 1]
```diff
@@ -13,22 +23,138 @@ type Config struct {
...
+	var commit *object.Commit
+	{
+		hash, err := repo.ResolveRevision(plumbing.Revision(ref))
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		commit, err = repo.CommitObject(*hash)
+		if err != nil {
+			log.Fatal(err)
```
Comment (TheoBrigitte) [inline pkg/gitrepo/repo.go]:
```suggestion
			return "", microerror.Mask(err)
```

File: pkg/gitrepo/repo.go [hunk 2]
```diff
@@ -13,22 +23,138 @@ type Config struct {
...
+			if ok {
+				lastTag = tag
+				break
+			}
+			c, err = c.Parent(0)
+			if errors.Is(err, object.ErrParentNotFound) {
+				lastTag = "v0.0.0"
+				break
+			} else if err != nil {
+				log.Fatal(err)
```
Comment (TheoBrigitte) [inline pkg/gitrepo/repo.go]:
```suggestion
				return "", microerror.Mask(err)
```

**Comment by kopiczko** [comment]
@TheoBrigitte is it fine now?

File: pkg/gitrepo/repo.go [hunk 3]
```diff
@@ -13,22 +22,138 @@ type Config struct {
...
+	}
+
+	fetchOpts := &git.FetchOptions{
+		Auth:  r.auth,
+		Force: true,
+	}
+
+	err = repo.Fetch(fetchOpts)
+	if errors.Is(err, git.NoErrAlreadyUpToDate) {
+		// Fall trough.
```
Comment (JosephSalisbury) [inline pkg/gitrepo/repo.go]:
```suggestion
		// Fall through.
```

**Comment by TheoBrigitte** [comment]
This doesn't work when you pass in a branch name. But we need this.

**Comment by kopiczko** [comment]
> This doesn't work when you pass in a branch name. But we need this.

It does when you prefix it with `origin/`. E.g. `origin/master`. We can add fallback here.

**Pull request merged**